src/cdk/tree/control/nested-tree-control.ts

```diff
@@ -0,0 +1,79 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {BaseTreeControl} from './base-tree-control';
+import {Observable} from 'rxjs/Observable';
+
+/** Nested tree control. Able to expand/collapse a subtree recursively for NestedNode type. */
+export class NestedTreeControl<T> extends BaseTreeControl<T> {
+
+  /** Construct with nested tree function getChildren. */
+  constructor(public getChildren: (node: T) => Observable<T[]>) {
+    super();
+  }
+
+  /**
+   * Expands all nodes in the tree.
+   *
+   * To make this working, the `nodes` of the TreeControl must be set correctly.
+   */
+  expandAll() {
+    this.expansionModel.clear();
+    this.nodes.forEach((node) => this._expandDescendants(node));
+    this.expansionChange.next(this.expansionModel.selected);
+  }
+
+  /** Gets a list of decedent nodes of a subtree rooted at given `node` recursively. */
+  getDescendants(node: T): T[] {
+    const descendants = [];
+    this._getDescendants(descendants, node);
+    return descendants;
+  }
+
+
+  /** Expands a subtree rooted at given `node` recursively. */
+  expandDescendants(node: T) {
+    this._expandDescendants(node);
+    this.expansionChange.next(this.expansionModel.selected);
+  }
+
+  /** Collapses a subtree rooted at given `node` recursively. */
+  collapseDescendants(node: T) {
+    this.expansionModel.deselect(node);
+    const subscription = this.getChildren(node).subscribe(children => {
+      if (children) {
+        children.forEach((child: T) => this.collapseDescendants(child));
+      }
+    });
+    subscription.unsubscribe();
+    this.expansionChange.next(this.expansionModel.selected);
+  }
+
+  /** Expands a subtree rooted at given `node` recursively without notification. */
+  protected _expandDescendants(node: T) {
+    this.expansionModel.select(node);
+    const subscription = this.getChildren(node).subscribe(children => {
+      if (children) {
+        children.forEach((child: T) => this.expandDescendants(child));
+      }
+    });
+    subscription.unsubscribe();
+  }
+
+  /** A helper function to get descendants recursively. */
+  protected _getDescendants(descendants: T[], node: T) {
+    descendants.push(node);
+    const subscription = this.getChildren(node).subscribe(children => {
+      if (children) {
+        children.forEach((child: T) => this._getDescendants(descendants, child));
+      }
+    });
+    subscription.unsubscribe();
```

Comment:
Rather than unsubscribing manually here, you could use the `first` operators instead.